### PR TITLE
feat(ec2-metadata-service): implement utils for ec2 metadata service (imds)

### PR DESCRIPTION
**Pull request by siddsriv**
### Issue
#4004 

### Description
Implementing utils for EC2 instance metadata service

### Testing
```console
  PASS  src/MetadataService.e2e.spec.ts
  MetadataService E2E Tests
    ✓ should fetch metadata token successfully (5 ms)
    ✓ should fetch metadata successfully with token (3 ms)
    ✓ should fetch metadata successfully (without token -- disableFetchToken set to true) (2 ms)
    ✓ should handle TimeoutError by falling back to IMDSv1 (2 ms)
    ✓ should handle specific error codes by falling back to IMDSv1 (6 ms)

Test Suites: 1 passed, 1 total
Tests:       5 passed, 5 total
Snapshots:   0 total
Time:        3.119 s, estimated 4 s
Ran all test suites.
Done in 3.66s.
```

### Additional context
Porting as-is from JSv2 to maintain parity.

### Checklist
- [x] If you wrote E2E tests, are they resilient to concurrent I/O?
- [x] If adding new public functions, did you add the `@public` tag and enable doc generation on the package?

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/ec2-metadata-service/src/MetadataService.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import { ENDPOINT_SELECTORS, IMDSv1_DISABLED_SELECTORS } from "./ConfigLoaders";
+import { MetadataServiceOptions } from "./MetadataServiceOptions";
+
+/**
+ * @public
+ */
+export class MetadataService {
+  private disableFetchToken: boolean;
+  private config: Promise<MetadataServiceOptions>;
+  /**
```
Comment (trivikr) [inline packages/ec2-metadata-service/src/MetadataService.ts:16]: nit: newline
```suggestion

  /**
```

File: packages/ec2-metadata-service/src/MetadataService.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        // If token fetch fails and IMDSv1 is not disabled, proceed without token (IMDSv1 fallback)
+      }
+    } // else, IMDSv1 fallback mode
+    const request = new HttpRequest({
+      method: options.method || "GET", // Default to GET if no method is specified
+      headers: headers,
+      hostname: endpointUrl.hostname,
+      path: endpointUrl.pathname + path,
+      protocol: endpointUrl.protocol,
+    });
```
Comment (trivikr) [inline packages/ec2-metadata-service/src/MetadataService.ts:67]: nit: newlines for readability
```suggestion

    const request = new HttpRequest({
      method: options.method || "GET", // Default to GET if no method is specified
      headers: headers,
      hostname: endpointUrl.hostname,
      path: endpointUrl.pathname + path,
      protocol: endpointUrl.protocol,
    });

```

File: packages/ec2-metadata-service/src/MetadataService.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    const handler = new NodeHttpHandler();
+    const endpointUrl = new URL(endpoint!);
+    const headers = options.headers || {};
+    /**
+     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
+     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+     */
+    if (this.disableFetchToken && ec2MetadataV1Disabled) {
+      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
+    }
```
Comment (trivikr) [inline packages/ec2-metadata-service/src/MetadataService.ts:45]: nit: newlines for readability
```suggestion

    /**
     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
     */
    if (this.disableFetchToken && ec2MetadataV1Disabled) {
      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
    }

```

File: packages/ec2-metadata-service/src/MetadataService.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    const endpointUrl = new URL(endpoint!);
+    const tokenRequest = new HttpRequest({
+      method: "PUT",
+      headers: {
+        "x-aws-ec2-metadata-token-ttl-seconds": "21600", // 6 hours;
+      },
+      hostname: endpointUrl.hostname,
+      path: "/latest/api/token",
+      protocol: endpointUrl.protocol,
+    });
```
Comment (trivikr) [inline packages/ec2-metadata-service/src/MetadataService.ts:96]: nit: newlines for readability
```suggestion

    const tokenRequest = new HttpRequest({
      method: "PUT",
      headers: {
        "x-aws-ec2-metadata-token-ttl-seconds": "21600", // 6 hours;
      },
      hostname: endpointUrl.hostname,
      path: "/latest/api/token",
      protocol: endpointUrl.protocol,
    });

```

**Review by trivikr — APPROVED** [review]
LGTM, just some nits on introducing newlines to improve readability.

File: packages/ec2-metadata-service/package.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    "@smithy/node-config-provider": "^2.2.1"
+  },
+  "devDependencies": {
+    "@tsconfig/recommended": "1.0.1",
+    "@types/node": "^14.14.31",
+    "concurrently": "7.0.0",
+    "downlevel-dts": "0.10.1",
+    "rimraf": "3.0.2",
+    "typescript": "~4.9.5",
+    "@aws-sdk/credential-providers": "3.533.0"
```
Comment (kuhe) [inline packages/ec2-metadata-service/package.json:39]: use `*` as the workspace version

**Review by kuhe — APPROVED** [review]
create a readme.md file

**Review by kuhe — APPROVED** [review]

**Pull request merged**